### ISSUE TITLE
Kinesis videostream only picked up in console about 1/10 times

### ISSUE THREAD

**Issue by GBoucher42** [awslabs/aws-sdk-android-samples#221]
I am running the video sample from an LG V20 Android device, running on Android 7.0. The application works, and sometimes I can actually see the video feed from the AWS console, but most of the time I have the No fragments found issue. I don't see what I'm doing differently when it works vs when it does not. I usually have this message, no matter hw long I wait. I've tried with front cam, back cam,

![image](https://user-images.githubusercontent.com/15200322/40280363-e7011cfe-5c1f-11e8-9680-4f526e08c2a0.png)

**And these are some of the messages appearing in my log while streaming on my device**

```
D/FrameUtility: frame timestamp: 1526829840641, index: 205, duration: 2, keyFrame: false, flags: 0
I/AndroidCameraMediaSource: updating sink with frame
D/KinesisVideoAndroidClient: PutFrame index: 205, pts: 15268298406410000, dts: 15268298406410000, duration: 20000, keyFrame: false, flags: 0
D/KinesisVideoAndroidClient: Data availability notification. Upload handle: 0, Size: 7024, Duration 20000 
D/KinesisVideoAndroidClient: Streamed 7024 bytes for stream digigrave with uploadHandle 0
D/EncoderWrapper: frame encoded | threadId=1
D/EncoderWrapper: encoding frame | threadId=1
D/EncoderWrapper: frame sent to encoder | threadId=1
D/EncoderWrapper: time between frames: 66ms
D/FrameUtility: frame timestamp: 1526829840706, index: 206, duration: 2, keyFrame: false, flags: 0
I/AndroidCameraMediaSource: updating sink with frame
D/KinesisVideoAndroidClient: PutFrame index: 206, pts: 15268298407060000, dts: 15268298407060000, duration: 20000, keyFrame: false, flags: 0
D/KinesisVideoAndroidClient: Data availability notification. Upload handle: 0, Size: 7468, Duration 20000 
D/KinesisVideoAndroidClient: Streamed 7468 bytes for stream digigrave with uploadHandle 0
D/EncoderWrapper: frame encoded | threadId=1
D/EncoderWrapper: encoding frame | threadId=1
D/EncoderWrapper: frame sent to encoder | threadId=1
D/EncoderWrapper: time between frames: 50ms
D/FrameUtility: frame timestamp: 1526829840756, index: 207, duration: 2, keyFrame: false, flags: 0
I/AndroidCameraMediaSource: updating sink with frame
D/KinesisVideoAndroidClient: PutFrame index: 207, pts: 15268298407560000, dts: 15268298407560000, duration: 20000, keyFrame: false, flags: 0
D/KinesisVideoAndroidClient: Data availability notification. Upload handle: 0, Size: 8499, Duration 20000 
D/KinesisVideoAndroidClient: Streamed 8499 bytes for stream digigrave with uploadHandle 0
D/EncoderWrapper: frame encoded | threadId=1
D/EncoderWrapper: encoding frame | threadId=1
D/EncoderWrapper: frame sent to encoder | threadId=1
D/EncoderWrapper: time between frames: 60ms
D/FrameUtility: frame timestamp: 1526829840817, index: 208, duration: 2, keyFrame: false, flags: 0
I/AndroidCameraMediaSource: updating sink with frame
D/KinesisVideoAndroidClient: PutFrame index: 208, pts: 15268298408170000, dts: 15268298408170000, duration: 20000, keyFrame: false, flags: 0
D/KinesisVideoAndroidClient: Data availability notification. Upload handle: 0, Size: 9637, Duration 20000 
D/KinesisVideoAndroidClient: Streamed 9637 bytes for stream digigrave with uploadHandle 0
D/EncoderWrapper: frame encoded | threadId=1
D/EncoderWrapper: encoding frame | threadId=1
D/EncoderWrapper: frame sent to encoder | threadId=1
D/EncoderWrapper: time between frames: 68ms
D/FrameUtility: frame timestamp: 1526829840884, index: 209, duration: 2, keyFrame: false, flags: 0
I/AndroidCameraMediaSource: updating sink with frame
D/KinesisVideoAndroidClient: PutFrame index: 209, pts: 15268298408840000, dts: 15268298408840000, duration: 20000, keyFrame: false, flags: 0
D/KinesisVideoAndroidClient: Data availability notification. Upload handle: 0, Size: 9274, Duration 20000 
D/KinesisVideoAndroidClient: Streamed 9274 bytes for stream digigrave with uploadHandle 0
D/EncoderWrapper: frame encoded | threadId=1
D/EncoderWrapper: encoding frame | threadId=1
D/EncoderWrapper: frame sent to encoder | threadId=1
D/EncoderWrapper: time between frames: 47ms
D/FrameUtility: frame timestamp: 1526829840931, index: 210, duration: 2, keyFrame: false, flags: 0
I/AndroidCameraMediaSource: updating sink with frame
D/KinesisVideoAndroidClient: PutFrame index: 210, pts: 15268298409310000, dts: 15268298409310000, duration: 20000, keyFrame: false, flags: 0
D/KinesisVideoAndroidClient: Data availability notification. Upload handle: 0, Size: 8087, Duration 20000 
D/KinesisVideoAndroidClient: Streamed 8087 bytes for stream digigrave with uploadHandle 0
D/EncoderWrapper: frame encoded | threadId=1
D/EncoderWrapper: encoding frame | threadId=1
D/EncoderWrapper: frame sent to encoder | threadId=1
D/EncoderWrapper: time between frames: 69ms
D/FrameUtility: frame timestamp: 1526829841000, index: 211, duration: 2, keyFrame: false, flags: 0
I/AndroidCameraMediaSource: updating sink with frame
D/KinesisVideoAndroidClient: PutFrame index: 211, pts: 15268298410000000, dts: 15268298410000000, duration: 20000, keyFrame: false, flags: 0
D/KinesisVideoAndroidClient: Data availability notification. Upload handle: 0, Size: 7194, Duration 20000 
D/KinesisVideoAndroidClient: Streamed 7194 bytes for stream digigrave with uploadHandle 0
D/EncoderWrapper: frame encoded | threadId=1
D/EncoderWrapper: encoding frame | threadId=1
D/EncoderWrapper: frame sent to encoder | threadId=1
D/EncoderWrapper: time between frames: 56ms
D/FrameUtility: frame timestamp: 1526829841056, index: 212, duration: 2, keyFrame: false, flags: 0
I/AndroidCameraMediaSource: updating sink with frame
D/KinesisVideoAndroidClient: PutFrame index: 212, pts: 15268298410560000, dts: 15268298410560000, duration: 20000, keyFrame: false, flags: 0
D/KinesisVideoAndroidClient: Data availability notification. Upload handle: 0, Size: 7410, Duration 20000 
D/KinesisVideoAndroidClient: Streamed 7410 bytes for stream digigrave with uploadHandle 0
D/EncoderWrapper: frame encoded | threadId=1
D/EncoderWrapper: encoding frame | threadId=1
D/EncoderWrapper: frame sent to encoder | threadId=1
D/EncoderWrapper: time between frames: 60ms
D/FrameUtility: frame timestamp: 1526829841116, index: 213, duration: 2, keyFrame: false, flags: 0
I/AndroidCameraMediaSource: updating sink with frame
D/KinesisVideoAndroidClient: PutFrame index: 213, pts: 15268298411160000, dts: 15268298411160000, duration: 20000, keyFrame: false, flags: 0
D/KinesisVideoAndroidClient: Data availability notification. Upload handle: 0, Size: 7268, Duration 20000 
D/KinesisVideoAndroidClient: Streamed 7268 bytes for stream digigrave with uploadHandle 0
D/EncoderWrapper: frame encoded | threadId=1
D/EncoderWrapper: encoding frame | threadId=1
D/EncoderWrapper: frame sent to encoder | threadId=1
D/EncoderWrapper: time between frames: 55ms
```

**Comment by cbommas** [comment]
hello @GBoucher42 

Sorry to hear you are running into issues. I will take a look at this and get back to you on this thread.

**Comment by MushMal** [comment]
@GBoucher42 hope we can help. It's not immediately clear from the logs whether the issue is on the producer side or the console. In order to shed some light, could I ask you to capture the logs from the moment you've started streaming so we can see whether there is a latency or bandwidth issue (generally, a networking issue).

**Comment by GBoucher42** [comment]
Ok, the log is quite long, but this is from running the app from Android Studio, streaming for about 20-30 seconds until I get the message that I get no fragments on my console, and pressing  "stop streaming"

[viode_producer_log.txt](https://github.com/awslabs/aws-sdk-android-samples/files/2029058/viode_producer_log.txt)

I've attached a file documents since the logs would not fit.

**Comment by MushMal** [comment]
Thanks for the logs

I see that your stream is around 18 fps and the observed upload rate is around 1.3 Mbps. Your network is able to keep it up with 0 latency (as in the not-yet-streamed buffer duration). I see the ACKs are streaming in - all looks very good.

At the end of the log I see the application being gracefully terminated.

I wonder if you could keep the stream running this particular instance and now view this in the console.

**Comment by GBoucher42** [comment]
I've leet it run for 3 minutes and still had the same issue, here are the logs. Sometimes it works, but rarely and for no apparent reason.
[video_producer_log.txt](https://github.com/awslabs/aws-sdk-android-samples/files/2029106/video_producer_log.txt)

**Comment by MushMal** [comment]
This is odd.. The second log also shows perfect streaming from the producer side. I wonder if there is an issue with the connectivity of the console. Can I ask you to do the following please?

* Specify what browser you are using.
* Attempt to use another browser (like Chrome, FireFox, etc..)
* Ensure you are trying to consume the right stream
* From the bottom-left, try to specify the server-side timestamps
* Try to specify the timeline that was playing "OK" and see if you replay just fine.

This will eliminate a number of issues so we can focus on the main issue.

**Comment by GBoucher42** [comment]
I am using Chrome, I have tried withwith firefox, same result.

I am sure that the stream is the right one as well. I've changed the name in the code so that it always populated the field with the right name.

The server-side timestamp is the same as my localtime (EDT). Which is also the same as the producer time stamp.

Yes, when I manage to stream, I can replay it afterwards.

I feel like it works more often when I reload the whole thing and relogin in, but I can't really create a working environnement.

**Comment by MushMal** [comment]
@GBoucher42 thanks for the info. So far, what I have is the following and please correct me if I am wrong:

* The producer logs indicate no issues at all with the streaming.
* The console playback is able to rollback and replay those fragments - consistently. If there were fragments then those will be played back consistently. If there were no fragments then they won't be..?

This sounds like you don't have any issues with the bandwidth on the console side (the consumer) and the logs I looked at just simply didn't catch the time when there was an error in streaming.

If this is the case then I would like to ask you to pinpoint the time when there is a problem in the console and include logs from a few seconds back which might include the time when there was an issue.

If this turns out to be some issue with the console, we could try another consumer - Java based real-time playback.

**Comment by MushMal** [comment]
You can modify the code of the sample java application to consume your own stream from the Android sample application. Check out the render sample from the parser library at:

https://github.com/aws/amazon-kinesis-video-streams-parser-library#example

You might also be able to adapt this for a playback on an Android device

**Comment by MushMal** [comment]
I am not sure I understand your scenario entirely. Are you going to consume the stream using a lambda?

The reason I asked to try a real-time consumer is to understand the following:

* Whether your producer has an issue streaming 
* Whether it's the transpackaging for the console playback
* Whether it's your browsers built-in MSE (playback engine)
* Whether it's fetching the index/getting the data from the storage that's failing on the backend.

The real-time path should help to eliminate those.

If this turns out to be the producer then we can try to capture the logs when the producer stops streaming so we can debug this issue on your end.

**Comment by GBoucher42** [comment]
Hi, I've managed to make it work with a certain hack. when I start the stream, if i lock my mobile (stops the stream) and unlock it (resumes the stream), then the console picks it up. I don't know if this helps.

Also, can I stream videos with sound? I don't see alot of documentation about this question.

**Comment by MushMal** [comment]
Interesting observation.. The demo application handles the Android lifecycle events so it resumes.

I am still quite not sure what's going on with the streaming intermittency - please try to get more data so we can help.

Regarding the audio. The processing is very similar to what the demo application does. We will try to add this functionality to the demo application soon. Basically, the idea is to create a new stream which is configured for AAC encoder specifics and enable streaming in parallel to the video. Stay tuned for that.

**Comment by GBoucher42** [comment]
What kind of data can I accumulate other tahn logs and working/failing scenarios? I will provide anything I can.

**Comment by MushMal** [comment]
The logs you've attached indicate perfect streaming for the period of time they were captured. I am just curious whether it's possible to catch the logs while your console playback is stuttering. Just trying to eliminate the producer side and ensure that the stream is being ingested without any interruptions.

**Comment by GBoucher42** [comment]
I've managed to change the code slightly and it works well.

In the following method from stream fragment.java, I've added the bolded line, which is called on the method onResume() which was called in my specific scenario. Now it works all the time.

    private void createClientAndStartStreaming(final SurfaceTexture previewTexture) {

        try {
            mKinesisVideoClient = KinesisVideoAndroidClientFactory.createKinesisVideoClient(
                    getActivity(),
                    Regions.US_EAST_1,
                    KinesisVideoDemoApp.getCredentialsProvider());

            mCameraMediaSource = (AndroidCameraMediaSource) mKinesisVideoClient
                    .createMediaSource(mStreamName, mConfiguration);

            mCameraMediaSource.setPreviewSurfaces(new Surface(previewTexture));

            **super.onResume();**
            resumeStreaming();
        } catch (final KinesisVideoException e) {
            Log.e(TAG, "unable to start streaming");
            throw new RuntimeException("unable to start streaming", e);
        }
    }

**Comment by MushMal** [comment]
Thanks a lot for the suggestion. We will evaluate this change and factor it in for the future improvements to the demo application

**Comment by raghavk92** [comment]
@GBoucher42  and @MushMal  
[log-kinesis.txt](https://github.com/awslabs/aws-sdk-android-samples/files/2057654/log-kinesis.txt)
i have a similar problem i have tried all the things mentioned here i think. I have tried super.onResume() that didnt work. What happens is that it takes more than 20 seconds till i see the video on Kinesis video streams in management console and then it after every 3-4 seconds it keeps showing loading for 10seconds then resumes. 

when i stop the stream and when i replay from the start of the stream i have to wait for the Pause button to become play so that i can replay  which also is a feels like bad UI or maybe i am missing something as in the management console because i cant navigate to the part where i started the stream  because i dont know at which time it started and i cant see a navigation bar..... It is a bit smoother while replaying but still it keeps showing loading for 7-8 seconds after every 4-5 seconds of video.

Sometimes I get fragment missing errors.

I have used aws versions 2.6.14,2.6.17,2.6.18  none working. And the latest version broke something so even the activity to start stream wasnt opening so had to revert to 2.6.14

I am attaching the logs for streaming on 2.6.14

Please help
Thanks

**Comment by minbi** [comment]
Hi @raghavk92 ,

We are aware of the issue with the latest 2.6.20 and working to fix the issue.

**Comment by raghavk92** [comment]
@minbi i had problems with streaming on 2.6.14  as mentioned above. The logs and details are for that. Is that normal?

**Comment by MushMal** [comment]
@raghavk92 log.txt indicates smooth streaming. Initially, your bandwidth was around 300Kbps which lead to accumulating a buffer over 4 seconds. Then, it increased to around 14Mbps so your buffer was streamed out just fine and caught up to the head (aka - the real-time).
The other one - log-kinesis.txt doesn't have enough data in there but there is no error or any issues I can see.

**Comment by muellerfr** [comment]
This issue has been closed because of inactivity. Please open a new issue in https://github.com/aws/aws-sdk-android/issues if you are still encountering problems.
